軽量標準Cライブラリ musl 環境下でビルドが通らない問題

Docker によるコンテナ化を試みてイメージ作成をしていたところ、ベースイメージが Alpine Linux 5.3 の場合にビルドできない問題に遭遇しました。調べて見たところ軽量化のために、標準Cライブラリに musl を使っているのが原因のようです。他にも Rasberry Pi の一部のディストリビューションや、組み込み環境で使われているようで、他の環境でも同様の問題が発生していると思われます。

## 再現方法
```
[recpt1@localhost:~]$ docker run --rm -it alpine:3.5 /bin/sh
/ # apk add --update build-base git autoconf automake
/ # mkdir -p /usr/local/src
/ # git clone https://github.com/stz2012/recpt1.git /usr/local/src/recpt1
/ # cd /usr/local/src/recpt1/recpt1
/usr/local/src/recpt1/recpt1 # ./autogen.sh
/usr/local/src/recpt1/recpt1 # ./autogen.sh ./configure
/usr/local/src/recpt1/recpt1 # make
```

```
revh="`git rev-list HEAD | wc -l 2> /dev/null`"; \
if [ -n "$revh" ] && [ "$revh" != "0" ] ; then \
	echo "const char *version = \"rev.$revh by stz2012\";" > version.h; \
else \
	echo "const char *version = \""c8688d7d6382_with_http_server_RC4 by stz2012"\";" > version.h; \
fi
gcc -MM recpt1.c decoder.c mkpath.c tssplitter_lite.c recpt1core.c recpt1ctl.c recpt1core.c -I../driver -Wall -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 > .deps
gcc -O2 -g -pthread -I../driver -Wall -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64  -c -o recpt1.o recpt1.c
In file included from recpt1.c:33:0:
recpt1core.h:53:16: error: redefinition of 'struct msgbuf'
 typedef struct msgbuf {
                ^~~~~~
In file included from recpt1.c:24:0:
/usr/include/sys/msg.h:42:8: note: originally defined here
 struct msgbuf {
        ^~~~~~
make: *** [<builtin>: recpt1.o] Error 1
```

--- a/recpt1/decoder.h
+++ b/recpt1/decoder.h
@@ -10,12 +10,12 @@
 #include <arib25/arib_std_b25.h>
 #include <arib25/b_cas_card.h>
 
-typedef struct decoder {
+typedef struct {
     ARIB_STD_B25 *b25;
     B_CAS_CARD *bcas;
 } decoder;
 
-typedef struct decoder_options {
+typedef struct {
     int round;
     int strip;
     int emm;
@@ -28,11 +28,11 @@ typedef struct {
     int32_t  size;
 } ARIB_STD_B25_BUFFER;
 
-typedef struct decoder {
+typedef struct {
     void *dummy;
 } decoder;
 
-typedef struct decoder_options {
+typedef struct {
     int round;
     int strip;
     int emm;

--- a/recpt1/recpt1.h
+++ b/recpt1/recpt1.h
@@ -15,12 +15,12 @@
 #define ISDB_T_NODE_LIMIT 24        // 32:ARIB limit 24:program maximum
 #define ISDB_T_SLOT_LIMIT 8
 
-typedef struct _BUFSZ {
+typedef struct {
     int size;
     u_char buffer[MAX_READ_SIZE];
 } BUFSZ;
 
-typedef struct _QUEUE_T {
+typedef struct {
     unsigned int in;        // 次に入れるインデックス
     unsigned int out;        // 次に出すインデックス
     unsigned int size;        // キューのサイズ
@@ -32,7 +32,7 @@ typedef struct _QUEUE_T {
     BUFSZ *buffer[1];    // バッファポインタ
 } QUEUE_T;
 
-typedef struct _ISDB_T_FREQ_CONV_TABLE {
+typedef struct {
     int set_freq;    // 実際にioctl()を行う値
     int type;        // チャンネルタイプ
     int add_freq;    // 追加する周波数(BS/CSの場合はスロット番号)

--- a/recpt1/recpt1core.h
+++ b/recpt1/recpt1core.h
@@ -45,17 +45,17 @@
 /* type definitions */
 typedef int boolean;
 
-typedef struct sock_data {
+typedef struct {
     int sfd;    /* socket fd */
     struct sockaddr_in addr;
 } sock_data;
 
-typedef struct msgbuf {
+typedef struct {
     long    mtype;
     char    mtext[MSGSZ];
 } message_buf;
 
-typedef struct thread_data {
+typedef struct {
     int tfd;    /* tuner fd */ //xxx variable
 
     int wfd;    /* output file fd */ //invariable

--- a/recpt1/tssplitter_lite.c
+++ b/recpt1/tssplitter_lite.c
@@ -24,6 +24,7 @@
 
 #include <fcntl.h>
 #include <sys/stat.h>
+#include <sys/types.h>
 #include "decoder.h"
 #include "recpt1.h"
 #include "tssplitter_lite.h"

